[SPMD] Error out SPMD mode

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54454 [SPMD] Error out SPMD mode**

According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.
3. Remove the assertion that prevents SPMD in DDP `join()` method, because now SPMD is already forbidden by the constructor. Also remove the relevant unit test `test_ddp_uneven_inputs_replicated_error`.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

Commit:
[SPMD] Error out SPMD mode

According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

zhaojuanmao commented:
should we change corresponding SPMD tests as well?

SciPioneer commented:
> should we change corresponding SPMD tests as well?

I updated test_c10d_spawn.py to fix one test, and I don't see any other test failure. Could you please point to me which SPMD test will still need a change at this time? Thanks!

zhaojuanmao commented on test/distributed/test_c10d_spawn.py:
```diff
@@ -241,11 +241,7 @@ def _test_base(self, net, inp, check_allclose=True):
-            # batch_size must be evenly divisible by num_gpus_used, take the largest one
-            num_gpus_used = [i for i in range(1, num_gpus + 1) if batch_size % i == 0][-1]
-            device_ids = list(range(num_gpus_used))
+            device_ids = [torch.cuda.current_device()]
```
I'm wondering whether we can check this test throws expected errors when using SPMD? after SPMD is fully killed, we can remove this test?

SciPioneer replied:
Added a few tests in test_c10d.py.

zhaojuanmao reviewed (commented):
did not notice you've changed test for test_c10d_spawn, added a comment. 

I searched a little bit, e.g., found out this test 'test_grad_layout_1devicemodule_2replicaperprocess' is also SPMD tests, but it was disabled because SPMD was broken before. possibly we can simply kill these tests after SPMD is fully retried.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -309,10 +309,8 @@ class DistributedDataParallel(Module):
-                   pass must be placed on the correct device. (default: all
-                   visible devices for single-device modules)
+                   CPU modules, ``device_ids`` must be ``None`` or an empty list.
+                   Multi-device module is no longer supported.
```
I don't think this is correct, we still do want to support multi-device modules right? If a module spans across multiple GPUs (ex: pipeline across 4 GPUs) we still want DDP to support that. I believe what we want to deprecate is the mode where a single DDP process is driving replicas across multiple GPUs.

SciPioneer replied:
Thanks for the explanation! Reverted the changes here.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -309,10 +309,8 @@ class DistributedDataParallel(Module):
                    only be provided when the input module resides on a single
                    CUDA device. For single-device modules, the i'th
                    :attr:`module` replica is placed on ``device_ids[i]``. For
```
Don't think these docs are accurate anymore since `device_ids` has to contain only one element right?

SciPioneer replied:
Updated. PTAL

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -388,6 +386,10 @@ def __init__(
         )
 
+        assert (
+            device_ids is None or len(device_ids) <= 1
```
I think we are mixing up two different modes here.

1. There is a multi-device model mode where the parameters of the module span across multiple GPUs (basically some form of model parallelism). We want DDP to support this mode and in this mode, device_ids will be None.
2. There is a single process multi device mode, where the model itself is on a single device, but len(device_ids) > 1 indicating that a single DDP process should replicate this model across multiple GPUs. We would like to deprecate this mode.

SciPioneer replied:
Thanks for the explanation! I totally ignored multi-device model mode here.

I think the assert condition is correct. Updated the error message here, to describe the above two modes separately.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -406,9 +408,9 @@ def __init__(
+            # Use the current device by default for single-device GPU modules
             if device_ids is None:
-                device_ids = _get_all_device_indices()
+                device_ids = [_get_current_device_index()]
```
We should document this behavior in our docs. Basically if device_ids is None and it is a single device module, it will use the current GPU device as the device for the model, so you need to use `torch.cuda.set_device` to set the appropriate device before you call DDP.

Another option would be to expect the user to place the model on the appropriate device for single device models (similar to how we expect this for multi-device models today). cc @mrshenli Thoughts?

SciPioneer replied:
> We should document this behavior in our docs. Basically if device_ids is None and it is a single device module, it will use the current GPU device as the device for the model, so you need to use `torch.cuda.set_device` to set the appropriate device before you call DDP.

Thanks for the suggestion! Added this requirement to the doctoring of DDP (in `device_ids` arg section).

> Another option would be to expect the user to place the model on the appropriate device for single device models (similar to how we expect this for multi-device models today).

In the pitch, it says "If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU." Therefore, I think `None` should be allowed for convenience.

zhaojuanmao replied:
vote for assuming user to place the model and input on the appropriate device for single device model when device_ids is none. This is consistent with multi-device models. also possibly better than assuming user has called torch.cuda.set_device()

pritamdamania87 commented on test/distributed/test_c10d_spawn.py:
```diff
@@ -241,11 +241,7 @@ def _test_base(self, net, inp, check_allclose=True):
-            # batch_size must be evenly divisible by num_gpus_used, take the largest one
-            num_gpus_used = [i for i in range(1, num_gpus + 1) if batch_size % i == 0][-1]
-            device_ids = list(range(num_gpus_used))
+            device_ids = [torch.cuda.current_device()]
```
We need a comprehensive set of tests that validate our new API contract, a few examples:

1. Specifying multiple device_ids results in an error.
2. Specifying device_ids None with a single device model works fine and uses the current device.
3. Specifying device_ids None with a multi device model works fine too.

SciPioneer replied:
Thanks for the suggestion! Added 3 test cases in test_c10d.py

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

mrshenli reviewed (commented):
cc @msbaines @blefaudeux are there any requirements from FAIRScale side that we should be aware of when retiring DDP SPMD mode?

blefaudeux commented:
@mrshenli on my side that's all good, I'll let @msbaines weigh in

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

msbaines commented:
> cc @msbaines @blefaudeux are there any requirements from FAIRScale side that we should be aware of when retiring DDP SPMD mode?

We don't use this DDP SPMD.

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -2441,6 +2442,27 @@ def _test_nccl_backend(
+    def test_nccl_backend_multi_device_ids_not_allowed(self):
+        int_devices = list(range(torch.cuda.device_count()))
+        devices = [torch.device("cuda:" + str(i)) for i in int_devices]
+        with self.assertRaises(AssertionError):
```
I think its probably better to raise a `ValueError` when a user passes invalid arguments to an API.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,19 @@ class DistributedDataParallel(Module):
+                   ``torch.cuda.set_device`` to set the appropriate device
+                   before calling DDP constructor.
+                   2) For multi-device modules and CPU modules,
+                   ``device_ids`` must be ``None`` or an empty list,
```
Why empty list, can we just keep it simpler by saying it should be None?

SciPioneer replied:
I didn't change the statement, and I don't know what's the motivation of allowing both `None` and an empty list here.

However, if we disallow an empty list, won't it break the backward compatibility?

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,19 @@ class DistributedDataParallel(Module):
+                   1) For single-device modules, ``device_ids`` should only
+                   contain exactly one device id, which represents the only
+                   CUDA device where the input module resides.
+                   Alternatively, if ``None`` is provided, the current device
+                   will be added to ``device_ids``. This requires calling
+                   ``torch.cuda.set_device`` to set the appropriate device
+                   before calling DDP constructor.
```
Lets go with Yanli's suggestion where if `None` is provided, we expect the user to place the model on the correct device instead of expecting `set_device` (this is consistent with multi-device models).

Also, avoid adding statements like 'the current device will be added to ``device_ids``', since that is implementation detail that the user shouldn't worry about.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,19 @@ class DistributedDataParallel(Module):
+                   before calling DDP constructor.
+                   2) For multi-device modules and CPU modules,
+                   ``device_ids`` must be ``None`` or an empty list,
+                   and input data for the forward pass must be placed on the
+                   correct device.
```
This statement is true for both points? So maybe just call it out after point 2) in a separate paragraph?

SciPioneer replied:
Thx for the suggestion! Updated.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -388,6 +393,14 @@ def __init__(
             "doesn't have any parameter that requires a gradient."
         )
 
+        assert (
```
Probably better to raise `ValueError` as I mentioned above. Looks like in other places we use `assert`, but I feel raising ValueError is more appropriate for invalid arguments. We could evaluate replacing the other asserts with ValueError as well in a separate PR. Although, it might not be possible to do that from a backward compatibility point of view since some user might be catching AssertionError and doing some error handling and if we change it to ValueError, it breaks their code. This is unlikely but possible.

SciPioneer replied:
Agree that `ValueError` is better in this case. Previously I tried to keep the consistency with the subsequent assertions. Changed to `ValueError`

Will replace other `AssertionError` with `ValueError` in a separate PR.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -388,6 +393,14 @@ def __init__(
 
+        assert (
+            device_ids is None or len(device_ids) <= 1
+        ), "device_ids can only be 1) None or an empty list,"
```
Isn't it simpler to just say device_ids should be None or just contain a single element?

SciPioneer replied:
Previously there is also a valid input which is an empty list. The error message is simplified now.

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,16 @@ class DistributedDataParallel(Module):
+                   ``device_ids`` must be ``None``.
+                   When ``device_ids`` is ``None`` for both cases,
+                   input data for the forward pass must be placed on the correct device.
+                   (default: all visible devices for single-device modules)
```
is this default still relevant? Should we set default to None? Otherwise, using single-device module on a multi-GPU machine (without setting `CUDA_VISIBLE_DEVICES`) will by default lead to crash?

rohan-varma replied:
What if we default to `torch.cuda.current_device()`? Seems like this would be a good default for single-device DDP module. Although, this might break multi-device DDP modules.

We should be careful when setting the default here that it doesn't break existing users for single and multi device.

SciPioneer replied:
Oh, forgot the change this line. I think it will be better to use `None` as the default value, which will be consistent with multi-module mode.

The drawback of using `torch.cuda.current_device()` is the requirement of calling `torch.cuda.set_device()` in advance.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -388,6 +390,9 @@ def __init__(
         )
 
+        if device_ids is not None and len(device_ids) > 1:
+            raise ValueError("device_ids can only be None or contain a single element.")
```
Could we add some more details here? i.e. Should we mention that SPMD in PyTorch 1.9+ is retired, and link to a GH issue for users to ping if they don't have a workaround for their use case?

SciPioneer replied:
I am afraid this will include too much details in the error message. Since we don't see any SPMD user at this time, we can keep the error message as simple as necessary at this time. If we see users ask questions about this in the future, then we can add more details by that time.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,16 @@ class DistributedDataParallel(Module):
+        device_ids (list of int or torch.device): CUDA devices.
+                   1) For single-device modules, ``device_ids`` should only
+                   contain exactly one device id, which represents the only
+                   CUDA device where the input module resides.
```
where the input module corresponding to this process resides?

SciPioneer replied:
Updated

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,16 @@ class DistributedDataParallel(Module):
-                   ``None`` or an empty list, and input data for the forward
-                   pass must be placed on the correct device. (default: all
-                   visible devices for single-device modules)
+        device_ids (list of int or torch.device): CUDA devices.
```
I wonder just for semantics, if it's worth providing a new `device` argument, so that new users don't have to wrap their head around the awkwardness of specifying only one device in a list of `device_ids` (though admittedly it's not much of an issue). Although, I'm guessing this will make the API too verbose to be feasible?

SciPioneer replied:
I think the existing API should be fine, as long as we view SPMD as a special case of a multi-device module, where the # of devices in a module is 1.

If we provide both `device` and `device_ids`, do you mean we shouldn't allow `device_ids` to contain a single element? If allowed, then the user can have two routes to control the same config -- what if the single id in `device` field and `device_ids` differ? This may also bring some unnecessary confusion.

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,16 @@ class DistributedDataParallel(Module):
-                   pass must be placed on the correct device. (default: all
-                   visible devices for single-device modules)
+        device_ids (list of int or torch.device): CUDA devices.
+                   1) For single-device modules, ``device_ids`` should only
```
nit: We should rephrase this: "``device_ids`` can contain ....". Since if we say "should only" contain, it seems like this is the only option, but below we say that it can be None as well.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,16 @@ class DistributedDataParallel(Module):
+                   1) For single-device modules, ``device_ids`` should only
+                   contain exactly one device id, which represents the only
+                   CUDA device where the input module corresponding to this process resides.
+                   Alternatively, ``device_ids`` can also be ``None``.
```
We should mention in this case, the user should place the module on the appropriate device.

SciPioneer replied:
Already mentioned after 2) to avoid duplication, because this is a common statement.

pritamdamania87 replied:
I think what we mentioned after 2) was that the input for the module should be placed on correct device. But we should also mention here that the actual module needs to be placed on the right device too.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,16 @@ class DistributedDataParallel(Module):
+                   Alternatively, ``device_ids`` can also be ``None``.
+                   2) For multi-device modules and CPU modules,
+                   ``device_ids`` must be ``None``.
+                   When ``device_ids`` is ``None`` for both cases,
```
nit: Maybe leave an empty line above after point 2).

SciPioneer replied:
Good to know that an empty line can be left in args docstring.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -396,7 +401,7 @@ def __init__(
         self.device_type = list(distinct_device_types)[0]
 
-        if self.device_type == "cpu" or self.is_multi_device_module:
+        if device_ids is None or self.device_type == "cpu" or self.is_multi_device_module:
```
Sorry I missed the case about backward compatibility for empty list, lets try to keep backward compatibility and add a condition here for empty list as well. Let's also add a unit test with empty list to ensure this contract is preserved. 

Its fine not to mention this in the docs though since its better that the user always passes in None.

SciPioneer replied:
Added a condition here as well as a unit test for SPSD mode.

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Test failure is real:

```
17:35:35 ======================================================================
17:35:35 ERROR [2.150s]: test_ddp_uneven_inputs_replicated_error (__main__.TestDistBackendWithSpawn)
17:35:35 ----------------------------------------------------------------------
17:35:35 Traceback (most recent call last):
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 335, in wrapper
17:35:35     self._join_processes(fn)
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 482, in _join_processes
17:35:35     self._check_return_codes(elapsed_time)
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 534, in _check_return_codes
17:35:35     raise RuntimeError(error)
17:35:35 RuntimeError: Process 0 exited with error code 10 and exception:
17:35:35 Traceback (most recent call last):
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 431, in run_test
17:35:35     getattr(self, test_name)()
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 337, in wrapper
17:35:35     fn()
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 102, in wrapper
17:35:35     return func(*args, **kwargs)
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 4487, in test_ddp_uneven_inputs_replicated_error
17:35:35     model.cuda(devices[0]), device_ids=devices, process_group=group
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/distributed.py", line 395, in __init__
17:35:35     raise ValueError("device_ids can only be None or contain a single element.")
17:35:35 ValueError: device_ids can only be None or contain a single element.
17:35:35 
17:35:35 Process 1 exited with error code 10 and exception:
17:35:35 Traceback (most recent call last):
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 431, in run_test
17:35:35     getattr(self, test_name)()
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 337, in wrapper
17:35:35     fn()
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 102, in wrapper
17:35:35     return func(*args, **kwargs)
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 4487, in test_ddp_uneven_inputs_replicated_error
17:35:35     model.cuda(devices[0]), device_ids=devices, process_group=group
17:35:35   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/distributed.py", line 395, in __init__
17:35:35     raise ValueError("device_ids can only be None or contain a single element.")
17:35:35 ValueError: device_ids can only be None or contain a single element.
```

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -2406,11 +2407,11 @@ def _test_gloo_backend(
     @requires_gloo()
     def test_gloo_backend_cpu_module_grad_is_view(self):
-        self._test_gloo_backend([torch.device("cpu")], [], gradient_as_bucket_view=True)
+        self._test_gloo_backend([torch.device("cpu")], None, gradient_as_bucket_view=True)
```
any reason for changing empty list to None? For BC, we need to confirm that empty list still works, no?

SciPioneer replied:
It seems the review suggestions have gone back and forth, from 1) keeping the empty list option for BC, 2) not keeping for the consistency with multi-device module, to 3) now keeping the empty list option for BC, but not documenting this option.

I have confirmed the empty list can still work, by keeping only a new unit test `test_nccl_backend_single_device_module_empty_device_ids`. Reverting these changes to the existing unit tests can still work, but since we don't want to mention this in the documentation, prefer changing the empty list to the advocated `None` for the most tests. After all, unit tests can also be viewed as a kind of documentation in some degree.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -305,14 +305,17 @@ class DistributedDataParallel(Module):
+                   CUDA device where the input module corresponding to this process resides.
+                   Alternatively, ``device_ids`` can also be ``None``.
+                   2) For multi-device modules and CPU modules,
+                   ``device_ids`` must be ``None``.
```
I see, we only allow None now. Any reason for this change?

SciPioneer replied:
@pritamdamania87 suggested only documenting supporting `None`, which is consistent with multi-device module, while still keep the backward compatibility underneath.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -396,7 +402,12 @@ def __init__(
-        if self.device_type == "cpu" or self.is_multi_device_module:
+        if (
+            device_ids is None
+            or len(device_ids) == 0  # For backward compatibility.
```
I see. As long as we still preserves BC, this change to `None` LGTM. But, let's leave at least one test that uses an empty list as `device_ids` to ensure test coverage.

SciPioneer replied:
I did create a new small unit test `test_nccl_backend_single_device_module_empty_device_ids` in test_c10d.py to cover this empty list case.

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Changes LGTM, but did you edit to fix the test failure on `test_ddp_uneven_inputs_replicated_error` on rocm? Looks like it still passes two devices to `device_ids`. 

This is skipped in our PR CIs as it requires 4 GPUs. Do we need to launch CI in case there might be other tests failing, or are we confident we addressed all?

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

SciPioneer commented:
> Changes LGTM, but did you edit to fix the test failure on `test_ddp_uneven_inputs_replicated_error` on rocm? Looks like it still passes two devices to `device_ids`.

Removed the assertion that prevents SPMD in DDP `join()` method, because now SPMD is already forbidden by the constructor. Also removed the relevant unit test `test_ddp_uneven_inputs_replicated_error`. Otherwise, this test will fail because the error message should be the one from the constructor, rather than from `join()` method.

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.
3. Remove the assertion that prevents SPMD in DDP `join()` method, because now SPMD is already forbidden by the constructor. Also remove the relevant unit test `test_ddp_uneven_inputs_replicated_error`.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.
3. Remove the assertion that prevents SPMD in DDP `join()` method, because now SPMD is already forbidden by the constructor. Also remove the relevant unit test `test_ddp_uneven_inputs_replicated_error`.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
There are a few conditional statements in python DDP class that's meant for SPMD mode, for example in `_sync_params` here: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L1263. Shall we remove these here as well or in follow up PRs?

pritamdamania87 reviewed (approved):
Looks good to me! Could you fix the small doc change I've mentioned inline before landing?

SciPioneer commented:
> There are a few conditional statements in python DDP class that's meant for SPMD mode, for example in `_sync_params` here: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L1263. Shall we remove these here as well or in follow up PRs?

I guess it's already disabled, so I could see a failure on that test. Will remove it in a separate PR.

Commit:
Update on "[SPMD] Error out SPMD mode"


According to the pitch in https://github.com/pytorch/pytorch/issues/47012

1. Let DDP error out if `device_ids` contains multiple devices.
2. If device_ids is not specified, DDP will use the provided model (module argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.
3. Remove the assertion that prevents SPMD in DDP `join()` method, because now SPMD is already forbidden by the constructor. Also remove the relevant unit test `test_ddp_uneven_inputs_replicated_error`.

#Closes: https://github.com/pytorch/pytorch/issues/47012

Differential Revision: [D27226092](https://our.internmc.facebook.com/intern/diff/D27226092/)

[ghstack-poisoned]
